shared/src/ReferenceCounter.cs

```diff
@@ -0,0 +1,217 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// A class that is used to handle an object that needs to be counted.
+    /// </summary>
+    /// <remarks>
+    /// Use of this class should be a static member variable inside of another class.
+    /// </remarks>
+    /// <example>
+    /// <code> public class ContainerUsage
+    /// {
+    ///     private ReferenceCounter&lt;object&gt; _refObject = new ReferenceCounter&lt;object&gt;();
+    ///     
+    ///     public ContainerUsage()
+    ///     {
+    ///         _refObject.Create(() => new object());
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    /// <typeparam name="T">Any class based object type.</typeparam>
+    public class ReferenceCounter<T> where T : class
+    {
+        private T _obejctToCount;
+
+        private object _lockObject = new object();
+
+        private Func<T, Task> _executeOnDispose;
+
+        private volatile int _referenceCount;
+
+        /// <summary>
+        /// The reference counted object.
+        /// </summary>
+        public T Value => _obejctToCount;
+
+        /// <summary>
+        /// The current count.
+        /// </summary>
+        public int Count => _referenceCount;
+
+        /// <summary>
+        /// Create the object or increment the counter.
+        /// </summary>
+        /// <remarks>
+        /// If the object is already set this will not execute the code in <paramref name="objectToCreate"/>. Instead it will return the object that was previously created.
+        /// 
+        /// 
+        /// 
+        /// This will return a mutable object. It is possible to alter the object state which could cause unexpected behavior.
```

Comment:
q - What kind of unexpected behavior are we referencing here?

Reply:
In general if you have an object that is reference counted there is no way to determine if the object has been changed. For example, an object could be disposed of while there is still a reference count of 1 but we'd have no way to tell and we'd still return the reference to the disposed object.